third_party/amd/lib/TritonAMDGPUTransforms/CanonicalizePointers.cpp

```diff
@@ -567,14 +567,19 @@ class ConvertAddPtrOp : public PointerCanonicalizationPattern<tt::AddPtrOp> {
                                            "tt.constancy"};
     SmallVector<NamedAttribute> propagatedAttrs =
         tt::filterDiscardableAttrs(addPtrOp.getOperation(), propagateList);
+    auto currPtrTy = llvm::dyn_cast<RankedTensorType>(addPtrOp.getType());
+    int currPtrRank = currPtrTy ? currPtrTy.getRank() : 1;
```

Comment:
instead of using `1` as a "do the default thing" indicator i would propose
```suggestion
    auto currPtrTy = llvm::dyn_cast<RankedTensorType>(addPtrOp.getType());
    auto doSetDiscardableAttrs = [](Operation* newAddPtrOp) {
      auto newPtrTy = llvm::dyn_cast<RankedTensorType>(newAddPtrOp.getType());
      // XOR to check scalar vs tensor 
      if (bool(newPtrTy) != bool(currPtrTy))
        return
      // both types are null (scalar) or ranked        
      if (!newPtrTy || (currPtrTy.getRank() == newPtrTy.getRank()))
        newAddPtrOp->setDiscardableAttrs(propagatedAttrs);          
    }
```

and then call below

Reply:
Shouldn't `newAddPtrOp.getType()` always return a scalar since it's adding to `fatPtrBase` which is always a scalar?

Reply:
AddPtrOp supports adding tensor offsets to scalar pointers. One of these branches is even checking if that's the current case that's being rewritten. I can't see the whole file right now (on my phone) to check whether the logic actually does force newPtr to be scalar but it's safer to just check.

Reply:
I've moved the common code into a lambda now but I prefer an explicit 1 for when the pointer is scalar to make it explicit that in that case we are doing a special case (between scalar and 1D)